[ClangImporter] define _GNU_SOURCE when importing glibc for Fuchsia.

zbowling opened the pull request:
Some functions and types are hidden behind feature flags.
_GNU_SOURCE exposes the most symbols and methods in glibc and musl.

(Part of https://github.com/apple/swift/pull/12955)

zbowling commented:
@jrose-apple I pulled this out of #12955

jrose-apple commented:
Thanks! I think it probably makes sense for this to just be a type-check test rather than an execution test. `%target-typecheck-verify-swift` is the thing to look for.

zbowling committed:
Define _GNU_SOURCE when importing glibc on Fuchisa

Some functions and types are hidden behind feature flags.

_GNU_SOURCE exposes the most symbols and methods in glibc
and musl. Currently only turning it on for the Fuchsia triple.

Also add a test for the missing extended C functions in the Glibc
module.

zbowling commented:
switched to that other test type. can you ping the ci bot to run tests for me?

CodaFi commented:
Serendipitously, @harlanhaskins and I just hit the issue this patch will fix.  Excellent work :sparks

CodaFi commented:
@swift-ci please test

CodaFi commented:
@swift-ci please test

gparker42 commented:
The Linux failure is caused by this change. Defining `_GNU_SOURCE` changes the signature of `strerror_r()`. Apparently POSIX and GNU defined incompatible things.

The Mac failure means you need to add `os(Fuschia)` support first, I think? (It is possible to run this PR's tests with another PR applied as well, if you have a separate PR with that change and you expect them all to get merged together.)

zbowling commented:
Yeah I figured the os(Fuchsia) would just emit a warning and not fail the tests on Mac until it landed (what it was doing locally). 

I wasn't seeing the strerror_r on my Ubuntu 14 machine. I'll do some investigating for a potential work around.

zbowling commented:
Oh the failure was in swiftpm, not swift. I didn't test that. 🙃 I'll submit a fix to swiftpm.

zbowling commented:
@gparker42 yeah I split this out #12955 which has all commits there. I was trying to land some of the changes separately so it wasn't such a big PR. Debating if that was a good idea now.

gparker42 commented:
Right. Might need to arrange some staging, with one or two smaller changes going in first to unblock the rest.

zbowling commented:
@gparker42 i added a change to swift PM to work around that one symbol.

Unfortunately I'm gating this on #if os(Linux) which means linux systems that use musl instead of glibc won't work (like arch linux). Since, while they support most of the same _GNU_SOURCE flags, the strerror_p oddity is not preserved. Android bionic also doesn't do that but at least that is treated as another OS. 

Something to solve maybe in a future evolution proposal to ease some of libc differences.

zbowling commented:
I can remove the #if os(Fuchsia) check in this commit and add it back in the Fuchsia one easy enough though.

gparker42 commented:
I suspect this now ought to go to swift-evolution, because `_GNU_SOURCE` has source-breaking effects other than merely providing more functions and types.

zbowling commented:
Yeah, I could gate the _GNU_SOURCE define so it's only applied to specific target triples (basically not Linux) until we settle this in evolution. That would for me to get Fuchsia up and running for now. 

I ran just ran through the rest of the glibc headers to see if there are any more hidden bombs like this one and it looks pretty good. Thankfully glibc uses a macro when they define two functions we the same name under different defines.

A lot of the paths I see from a quick grep are not super concerning because most of them involve LARGEFILE64 defines that weren't setting before or even now and totally moot on 64bit anyways. A few __REDIRECT cases are where there are differences between BSD_SOURCE and GNU_SOURCE emitted functions but we weren't emitting a function for either one today because we didn't define either one (specifically linux's signal() vs sysv's signal()). 

The only breakage I can see possibly is in vfwscanf, vwscanf, vswscanf, fwscanf, wscanf, and swscanf where they accept different scan args under GNU_SOURCE  but only one is a conflicting arg (but also it's wchar and I feel bad if you are attempting to use wchar in 2017 and especially in swift for anything).

zbowling commented:
Yeah, I did a diff from the generated imported module with the ide test tool. The only signature difference seems to be only the ```strerror_r()``` function. (The former lead maintainer of glibc for the past 30 years sits next to me now so I might bug him about what is up with this weird function). 

Interestingly wchar.h isn't even in the glibc.modulemap so the vfwscanf, vwscanf, vswscanf, fwscanf, wscanf, and swscanf functions are not actually an issue.

zbowling commented:
@jrose-apple @gparker42 issue came up here perviously: https://github.com/apple/swift-package-manager/pull/721

zbowling commented:
@gparker42 @jrose-apple I have a better idea. We could also set something like this instead which includes everything in all the POSIX/C/etc standards without the enabling the GNU specific extensions.

```
"-D_ATFILE_SOURCE=200809L",
"-D_LARGEFILE64_SOURCE=200809L",
"-D_ISOC99_SOURCE=200809L",
"-D_XOPEN_SOURCE_EXTENDED=200809L",
"-D_POSIX_SOURCE=200809L",
"-D_POSIX_C_SOURCE=200809L",
"-D_POSIX_SOURCE=700"
```

This enable a good number of the new functions without hitting the strerror_r() rewrite. We still miss out on common/good GNU extension functions like ```dup3```, ```pipe2```, and ```qsort_r``` but if we need those we can manually wrap them as needed with some c code in the platform folder and manually expose in the glibc module.

harlanhaskins commented:
Does that still block `dladdr` and the like, though? IIRC it's guarded by `__USE_GNU` specifically.

zbowling commented:
Yeah, dladdr will not show up without defining _GNU_SOURCE explicitly. 

Well my only other crazy idea is we just rewrite stderror_r to something else obfuscated with an APINotes hack and then make a manual wrapper in C that we manually expose to swift in the glibc module that calls the non-GNU extension version.

jrose-apple commented:
I feel like this really does need more discussion at this point from more of the Linux people. This should go to swift-evolution or swift-dev. Meanwhile, I think adding the most limited set of feature macros Fuchsia needs, guarded by a target OS check, is a reasonable way to move forward.

zbowling commented:
Sounds good.

zbowling commented:
@jrose-apple @gparker42 I wrapped this in a triple check for Fuchsia for now. 

I kept the test since FreeBSD and Android will still expose the functions used in the test even without the feature macro defined in their libcs. For now I just marked the test to exclude linux for now until we figure out what we want to do.

phausler commented:
@swift-CI please test

zbowling commented:
err my force push broke the test CI it looks like

phausler commented:
looks like it is just the first pass, the second one that is running now should be the right sha

gparker42 commented:
That CI hiccup is normal. As described in the failure log it has automatically started another CI run.

jrose-apple commented:
I was wondering if you don't want to define all of `_GNU_SOURCE` on Fuchsia yet in case the conclusion is to define something more restrictive, but it's probably fine.

zbowling commented:
I think I can go lower than _GNU_SOURCE on Fuchsia but I need test. Our libc is forked from musl and musl doesn't go to the extent that glibc goes with feature macros (musl strives to be compatible with glibc for that flag in a lot of cases but isn't perfect) so I need to verify something like _XOPEN_SOURCE + _POSIX_SOURCE will turn on everything we have already using in our swift code internally.

zbowling commented:
I switched to "xfail: linux" instead of "unsupported: os=linux" because that seems to still fail the linux test on the CI.

gparker42 commented:
`OS=Linux` is not a supported test. `XFAIL: linux` should work, although that might also XFAIL the Android build too.

zbowling commented:
I rebased/squashed this onto master. This should be good to test.

gottesmm commented:
@swift-ci please test

gottesmm commented:
@swift-ci please test

zbowling commented:
merge?

jrose-apple commented on test/stdlib/LibcFeatureFlag.swift:
```diff
@@ -0,0 +1,33 @@
+// RUN: %target-typecheck-verify-swift
+// REQUIRES: executable_test
```
Since we're not actually executing the test, there's no need for this anymore…

zbowling replied:
I can kill it but I was hoping to leave it document the inconsistency/differences with glibc coverage depending on the platform you run the test on (like this test will pass on FreeBSD and Android but not Linux) until we can figure out what feature flag we want to set. Once we add a feature flag for Linux this test should start passing and we can kill the XFAIL :smile:

jrose-apple replied:
:-/ "REQUIRES: executable_test" doesn't distinguish between Linux and FreeBSD, though.

zbowling replied:
Gotcha. I'll kill the test until we figure out how to enable this flag for Linux eventually.

gparker42 replied:
As I understand it you want something like `REQUIRES: executable_test` and `UNSUPPORTED: OS=linux-gnu`.

gparker42 replied:
(`XFAIL` always runs the test, but expects the test to fail on the specified configurations. `UNSUPPORTED` skips the test on the specified configurations.)

jrose-apple commented on test/stdlib/LibcFeatureFlag.swift:
```diff
@@ -0,0 +1,33 @@
+// drand48,mrand48,etc are only visible if _XOPEN_SOURCE (or better) is defined
+// on Glibc
+let a = drand48()
+print(a)
```
…nor do we need to actually print anything. `_ = drand48()` is sufficient.

zbowling replied:
yeah I wasn't it mattered since it was a %target-typecheck-verify-swift test. do we actually execute those?

jrose-apple replied:
We don't (deliberately, to speed up testing time). Might as well not include things that aren't being tested.

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -533,6 +533,17 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
       });
     }
   } else {
+    // Idealy we should turn this on for all Glibc targets. This will have some
```
Typo: "ideally"

zbowling replied:
Oops

jrose-apple reviewed (commented):
I'm good with the content at this point. A few suggestions still on the test (sorry to push this through another test cycle).

zbowling commented:
Killed the test and reworded the comment (with the spelling fix)

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
Merged. Thanks, Zac.
